common/src/main/java/org/apache/nemo/common/ir/IRDAG.java

```diff
@@ -0,0 +1,286 @@
+    final DAGBuilder builder = new DAGBuilder();
+
+    // Insert the vertex.
+    builder.addVertex(messageBarrierVertex);
+
+    // Current metric collection id.
+    final int currentMetricCollectionId = metricCollectionId.incrementAndGet();
+
+    dag.topologicalDo(v -> {
+      // None of the existing vertices are deleted.
+      builder.addVertex(v);
+
+      for (final IREdge edge : dag.getIncomingEdgesOf(v)) {
+        if (edge.equals(edgeToGetStatisticsOf)) {
+          // MATCH!
+          builder.addVertex(messageBarrierVertex);
+          builder.addVertex(messageAggregationVertex);
+
+          // Clone the edgeToGetStatisticsOf
+          final IREdge clone = new IREdge(
+            CommunicationPatternProperty.Value.OneToOne, edge.getSrc(), messageBarrierVertex);
+          clone.setProperty(EncoderProperty.of(edge.getPropertyValue(EncoderProperty.class).get()));
+          clone.setProperty(DecoderProperty.of(edge.getPropertyValue(DecoderProperty.class).get()));
+          edge.getPropertyValue(AdditionalOutputTagProperty.class).ifPresent(tag -> {
+            clone.setProperty(AdditionalOutputTagProperty.of(tag));
+          });
+          builder.connectVertices(clone);
+
+          // messageBarrierVertex to the messageAggregationVertex
+          final IREdge edgeToABV = edgeBetweenMessageVertices(messageBarrierVertex,
+            messageAggregationVertex, mbvOutputEncoder, mbvOutputDecoder, currentMetricCollectionId);
+          builder.connectVertices(edgeToABV);
+
+          // Connection vertex
+          // Add a control dependency (no output)
+          final IREdge emptyEdge =
+            new IREdge(CommunicationPatternProperty.Value.BroadCast, messageAggregationVertex, v);
+          builder.connectVertices(emptyEdge);
+
+          // The original edge
+          // We then insert the vertex with MessageBarrierTransform and vertex with MessageAggregateTransform
+          // between the vertex and incoming vertices.
+          final IREdge edgeToOriginalDst =
+            new IREdge(edge.getPropertyValue(CommunicationPatternProperty.class).get(), edge.getSrc(), v);
+          edge.copyExecutionPropertiesTo(edgeToOriginalDst);
+          edgeToOriginalDst.setPropertyPermanently(MetricCollectionProperty.of(currentMetricCollectionId));
+          builder.connectVertices(edgeToOriginalDst);
+        } else {
+          // NO MATCH, so simply connect vertices as before.
+          builder.connectVertices(edge);
+        }
+      }
+    });
+
+    dag = builder.build(); // update the DAG.
+  }
+
+  ////////////////////////////////////////////////// "Unsafe" direct reshaping (semantic-preserving is not guaranteed).
+
+  /**
+   * "Unsafe" direct reshaping (semantic-preserving is not guaranteed).
+   * @param unsafeReshaping a function that directly reshapes the underlying DAG.
+   */
+  public void unSafeDirectReshaping(final Function<DAG<IRVertex, IREdge>, DAG<IRVertex, IREdge>> unsafeReshaping) {
```

Comment:
Maybe something like 'reshapeUnsafely' or 'applyUnsafeDirectReshaping' or something like that? Unsafe is a single word by the way. I think it would be better to keep the names of method that perform a specific action as verbs. Could you double check on this throughout your PR?

Reply:
DAG<IRVertex, IREdge> seems to be still used here. I don't see a reason for using both forms.